torch/_higher_order_ops/flex_attention.py

```diff
@@ -70,6 +71,114 @@ def _permute_strides(out: torch.Tensor, query_strides: Tuple[int, ...]) -> torch
     return new_out
 
 
+@torch.library.custom_op("mylib::zeros_and_scatter", mutates_args=())
+def zeros_and_scatter(
+    shape: List[int],
+    indices: List[Tensor],
+    indices_dims: List[int],
+    vals: Tensor,
+    rightmost_dim: int,
+) -> Tensor:
+    grad = torch.zeros(shape, device=vals.device, dtype=vals.dtype)
+
+    # For multiple indexes, it's possible the indices are not ordered,
+    # so we have to transpose the tensor before the index_put.
+    # Meanwhile, it's possible the indices are not continuous, e.g,
+    # bias[b, kv_idx, q_idx] whose indices_dims are [0, 3, 2].
+    if len(indices) > 1:
+        new_indices_dims = list(range(4))
+        ordered_indices_dims = sorted(indices_dims)
+        for i, idx in enumerate(ordered_indices_dims):
+            new_indices_dims[idx] = indices_dims[i]
+        vals = torch.permute(vals, new_indices_dims)
+        indices = list(torch.meshgrid(indices, indexing="ij"))
+
+    # We need to unsqueeze the indices to facilitate the broadcasting in the next step.
+    leftmost_dim = min(indices_dims)
+    if rightmost_dim - leftmost_dim + 1 == len(indices):
+        offset = -1
+        for i in range(3, leftmost_dim, -1):
+            if i in indices_dims:
+                offset = offset - 1
+            else:
+                for j, idx in enumerate(indices):
+                    indices[j] = idx.unsqueeze(offset)
+
+    # Broadcast the indices to match the shape of vals.
+    for i, idx in enumerate(indices):
+        indices[i] = idx.expand(vals.shape)
```

Comment:
All the code above is going to handle different bias variants + indexing, it supports all the cases listed in https://gist.github.com/yanboliang/af0e6a6b9dedb2742fd45cd5d017d2a6, but probably there is still gap. But it has better coverage than eager SDPA.